lib/LaTeXML/Core/Document.pm

```diff
@@ -1082,17 +1087,39 @@ sub openText_internal {
   my $qname;
   if ($$self{node}->nodeType == XML_TEXT_NODE) {    # current node already is a text node.
     print STDERR "Appending text \"$text\" to " . Stringify($$self{node}) . "\n" if $LaTeXML::Core::Document::DEBUG;
+    my $parent = $$self{node}->parentNode;
+    if ($LaTeXML::BOX && $parent->getAttribute('_autoopened')) {
+      $self->appendTextBox($parent, $LaTeXML::BOX); }
     $$self{node}->appendData($text); }
   elsif (($text =~ /\S/)                            # If non space
     || $self->canContain($$self{node}, '#PCDATA')) {    # or text allowed here
     my $point = $self->find_insertion_point('#PCDATA');
     my $node  = $$self{document}->createTextNode($text);
+    if ($point->getAttribute('_autoopened')) {
+      $self->appendTextBox($point, $LaTeXML::BOX); }
     print STDERR "Inserting text node for \"$text\" into " . Stringify($point) . "\n"
       if $LaTeXML::Core::Document::DEBUG;
     $point->appendChild($node);
     $self->setNode($node); }
   return $$self{node}; }                                # return the text node (current)
 
+# Since xml text nodes don't have attributes to record the origining box,
+# we need to manage the accumulation of autoOpen'ed boxes
+# Indeed, propogate it to ancestors if they were autoOpened for same cause (box)
+sub appendTextBox {
```

Comment:
Very interesting technique, clever too. Properly tracking the boxes for auto-opened nodes is super useful, wish I had realized it was missing.

Reply:
Any thoughts about tracking this better on the generic level in latexml? Or is the need to bookkeep specific to the pgf binding? I'm still absorbing the fine details.

Reply:
It eventually did get more generic, as I had to add a similar box patch-up when adding xml elements to auto-opened nodes.  It is also generic in that it applies to all xml, but it was driven by pgf being the application that depends most on having the right boxes attached for sizing estimates.